### ISSUE TITLE
libc++ should (behind a macro that can enable and disable) add [[nodiscard]] unique_ptr::release

### ISSUE THREAD

**Issue by chandlerc** [llvm/llvm-project#30246]
|  |  |
| --- | --- |
| Bugzilla Link | [30898](https://llvm.org/bz30898) |
| Version | unspecified |
| OS | All |
| CC | @mclow |

## Extended Description 
We have a C++ annotation ([[nodiscard]]) which allows warnings when returned values are ignored. This should be possible to enable on std::unique_ptr<...>::release()'s returned pointer so that we don't accidentally leak memory.

**Comment by ldionne** [comment]
As explained in https://github.com/llvm/llvm-project/pull/110847, it turns out that adding `[[nodiscard]]` on `unique_ptr::release()` seems to trigger a lot of false positives. I'm not convinced this is a win.

Closing.

**Comment by philnik777** [comment]
Reopening, since I'd like us to investigate applying `[[nodiscard]]` only when using the default deleter. I don't know whether that's fruitful, but I think it's certainly worth a shot, since memory leaks are at the very least annoying to track down. If we could get at least the default deleter case to warn with a high signal to noise ratio that would already be a huge win I think.

**Comment by ldionne** [comment]
My thinking is that the main reason for so much noise is not the presence of a custom deleter, but the fact that people often do things like

```
std::unique_ptr<T> ptr = ...;
use(ptr);
reclaim(ptr.get());
ptr.release();
```

I would expect that we'll see just as much noise (as reported in #110847) if we only add the attribute when we have a default deleter. But I'm open to investigating, if we find that it's not the case then this is probably worth enabling.
